training/CHAPEL_TUTORIAL.md

```diff
@@ -0,0 +1,861 @@
+[`toUpper`](https://chapel-lang.org/docs/language/spec/strings.html?highlight=totitle#String.string.toUpper)
+and [`join`](https://chapel-lang.org/docs/language/spec/strings.html?highlight=totitle#String.string.join)!
+
+
+<a id="TIY3"></a>
+<a id="TIY3"></a>
+#### Try It Yourself 3: `title`
+Problem:
+Use Aggregation to transform the SegString from [example 15](#ex15) into title case.
+Be sure to use the chapel builtin [`toTitle`](https://chapel-lang.org/docs/language/spec/strings.html?highlight=totitle#String.string.toTitle)
+and `getLengths` from example 15.
+<details>
+  <summary>Potential Solution</summary>
+
+```Chapel
+use CopyAggregation;
+
+proc title() throws {
+  var titleVals: [values.domain] string;
+  const lengths = getLengths();
+  forall (off, len) in zip(offsets, lengths) with (var valAgg = new DstAggregator(string)) {
+    var i = 0;
+    for char in ''.join(values[off..#len]).toTitle() {
+      valAgg.copy(titleVals[off+i], char);
+      i += 1;
+    }
+  }
+  return titleVals;
+}
+writeln(title());
+```
+</details>
+Expected Output:
+
+```console
+S s s 0  S s 1  S 2
+```
+
+<a id="filter"></a>
+## Filtering
+<a id="ex17"></a>
+#### Example 17: Filtering
+We can filter an iterator to only operate on values matching a certain condition:
+```chapel
+writeln([i in 0..#10] if i%2 == 0 then -i);
+```
+```console
+0 -2 -4 -6 -8
+```
+
+<a id="TIY4"></a>
+#### Try It Yourself 4: Sum Odd Perfect Squares <=25
+Problem:
+Use filtering and reduce to sum all odd perfect squares less than or equal to `25`
+<details>
+  <summary>Potential Solution</summary>
+
+```Chapel
+writeln(+ reduce [i in 0..5] if i%2 != 0 then i**2);
+```
+</details>
+Expected Output:
+
+```console
+35
+```
+
+<a id="bool_expand_and_compress"></a>
+## Boolean Compression and Expansion Indexing
+[Boolean compression and expansion indexing](https://github.com/Bears-R-Us/ArkoudaWeeklyCall/blob/main/Indexing/ArkoudaBooleanIndexing.pdf)
+is an application of `reduce scan op`, `forall`, and filtering. It was covered in an [Arkouda Weekly Call](https://github.com/Bears-R-Us/ArkoudaWeeklyCall).
+
+<a id="ex18"></a>
+#### Example 18: Boolean Compression Indexing
+Compression indexing is reducing an array to only the values meeting a certain condition.
+This is a common operation in numpy.
+
+Compression indexing in `numpy`:
+```python
+>>> X = np.array([1, 2, 5, 5, 1, 5, 2, 5, 3, 1])
+>>> Y = X[X==5]
+>>> Y
+array([5, 5, 5, 5])
+```
+We can accomplish the same result in Chapel using tools from this guide! Specifically `reduce scan op`, `forall`, and filtering.
+```chapel
+var X = [1, 2, 5, 5, 1, 5, 2, 5, 3, 1];
+writeln("X = ", X, "\n");
+
+// we begin by creating a boolean array, `truth`, indicating where the condition is met
+var truth = X == 5;
+writeln("truth = ", truth);
+
+// we use `truth` to create the indices, `iv`, into the compressed array
+// note: `iv[truth] = [0, 1, 2, 3]`
+var iv = + scan truth - truth;
```

Comment:
I was wondering, when you were introducing scans, whether you should define inclusive vs. exclusive scans and point out that Chapel's are inclusive.  This made me wonder again, since I think the `- truth;` is essentially changing the inclusive result into an exclusive scan.  Note that we'd like for Chapel to support exclusive scans as well, but it's never been prioritized...